### PR TITLE
feat: config: Make fast path re-elect configurable

### LINKED ISSUE #1511: Leader survival across restart may cause state machine inconsistency for external readers
### Description

OpenRaft provides a feature that allows a leader to survive a service restart. When a node that was a leader before restart comes back quickly enough (before any other node triggers an election), it can continue serving as leader without re-election.

This differs from standard Raft, which requires a restarted node to either wait for election timeout or participate in a new election.

### Problem

While this feature improves availability, it introduces a potential consistency issue:

1. **State machine state loss**: When the node restarts, the state machine may lose in-memory state that wasn't persisted to disk. It needs to reapply logs to restore state.

2. **Committed index not guaranteed persisted**: The `committed` index is not a required persistence item for user applications. If the application doesn't persist it correctly, after restart the state machine may only be partially restored.

3. **Silent inconsistency**: External readers that query the state machine may see different results than before the restart, because:
   - The state machine is in a "half-restored" state
   - The leader continues serving requests without the application being aware of the incomplete state
   - There's no mechanism for users to detect this inconsistency

### Scenario

1. Node A is leader with committed index at 100
2. State machine has applied up to log 100, but some state exists only in memory
3. Node A restarts quickly
4. Node A resumes as leader (no re-election occurred)
5. State machine reapplies logs but user application didn't persist committed index
6. External reader queries and gets inconsistent/incomplete data
7. User has no way to detect this problem

### Solution

The solution is straightforward: **user applications must guarantee that `save_committed()` in their `RaftLogStorage` implementation persists the committed index before returning**.

If this guarantee is met, the state machine will always restore to the exact same state after restart, and no inconsistency will occur.

### Action Items

- [ ] Document this requirement clearly for users who rely on leader survival across restart
- [ ] Consider adding a configuration flag or warning when this feature is enabled

### PR DESCRIPTION
**Pull request by jkosh44**
Previously, when a leader restarted fast enough, it would re-elect itself as a leader without an election. This behavior improves availability but differs from standard Raft and introduces potential inconsistencies.

This change adds a configuration that allows users to disable this behavior.

Fixes #1511

**Checklist**

- [X] Updated guide with pertinent info (may not always apply). 
- [X] Squash down commits to one or two logical commits which clearly describe the work you've done.
- [X] Unittest is a friend:)

- - -
This change is [Reviewable](https://reviewable.io/reviews/databendlabs/openraft/1771)

### CODE REVIEW

**Comment by jkosh44** [comment]
I'm very open to bike shedding on anything, in particular:
- The config name
- The config placement

Also please let me know if the added documentation is accurate and sufficient.

File: openraft/src/config/config.rs [hunk 1]
```diff
@@ -431,6 +433,45 @@ pub struct Config {
...
+    /// When the above conditions are met and this setting is disabled, then a restarted leader can
+    /// learn the current commit log id by participating in a new election or receiving a message
+    /// from the new leader.
+    ///
+    /// See: [`docs::data::log_pointers`].
+    ///
+    /// [`docs::data::log_pointers`]: `crate::docs::data::log_pointers#optionally-persisted-committed`
+    // clap 4 requires `num_args = 0..=1`, or it complains about missing arg error
+    // https://github.com/clap-rs/clap/discussions/4374
+    #[since(version = "0.11.0")]
```
Comment (jkosh44) [inline openraft/src/config/config.rs]: I wasn't sure what value to put here.

**Comment by drmingdrmer** [comment]
Thanks @jkosh44 — right problem to target, and gating the startup leader-restore is the natural first move. But I don't think it's complete yet, because Openraft persists leadership differently from standard Raft.

Standard Raft persists only `(currentTerm, votedFor)`, so a restarted leader can simply come back as a follower — nothing on disk says "I'm the leader." Openraft instead records leadership in the **committed `Vote`** (`is_leader = is_leading && vote.is_committed()`). With the flag off, the PR skips the restore branch and the node becomes Follower/Learner, but it leaves that committed vote in place. So after `startup()`:

- `server_state == Follower`, `leader_ref() == None`
- yet `state.is_leader(&id)` still returns `true` (the committed self-vote is untouched)
- and storage still holds that committed vote

That's a "follower" that `is_leader()` reports as a leader — an in-memory contradiction — on top of a memory/storage divergence. The new test actually shows it: the committed vote is left as-is and `take_commands()` is empty, so nothing reconciles or persists the change.

So emulating standard Raft here means more than skipping the leader-state restore — we also have to reconcile the persisted leader info. The direction I lean toward: when the node reverts to follower instead of restoring leadership, demote the committed vote to uncommitted (keep the term, drop the committed bit so `is_leader()` goes false) and persist it with a `SaveVote`, leaving a clean follower at the same term. Let me think through the edge cases (lease handling, which term we keep) before we settle the exact handling.

On the config:
- **Type:** make it `Option<bool>` defaulting to `None`, resolved at runtime rather than baked into a `bool` — mirroring `allow_log_reversion`. `None` lets us tell "user didn't choose" apart from an explicit `Some(true)`/`Some(false)` even when the effective value is the same, so we can warn when the consistency-sensitive default is in effect and evolve the default later without silently changing behavior.
- **Name:** `enable_leader_restore`. "reelect" is a little misleading — this path runs *no* election; the node resumes from its persisted committed vote — whereas "restore" matches the `startup()` code ("restore it as leader") and the `enable_*` convention (`enable_elect`, etc.).

Placement in `Config` is fine. Minor: `#[since]` should be `0.10.0` (the next release line), not `0.11.0`.

**Comment by jkosh44** [comment]
Thank you for the review! I just pushed a new commit.

> On the config:
> 
>    - **Type:** make it `Option<bool>` defaulting to `None`, resolved at runtime rather than baked into a `bool` — mirroring `allow_log_reversion`. `None` lets us tell "user didn't choose" apart from an explicit `Some(true)`/`Some(false)` even when the effective value is the same, so we can warn when the consistency-sensitive default is in effect and evolve the default later without silently changing behavior.
> 
>    - **Name:** `enable_leader_restore`. "reelect" is a little misleading — this path runs _no_ election; the node resumes from its persisted committed vote — whereas "restore" matches the `startup()` code ("restore it as leader") and the `enable_*` convention (`enable_elect`, etc.).
> 
> 
> Placement in `Config` is fine. Minor: `#[since]` should be `0.10.0` (the next release line), not `0.11.0`.

These have all been addressed and updated.

> Standard Raft persists only `(currentTerm, votedFor)`, so a restarted leader can simply come back as a follower — nothing on disk says "I'm the leader." Openraft instead records leadership in the **committed `Vote`** (`is_leader = is_leading && vote.is_committed()`). With the flag off, the PR skips the restore branch and the node becomes Follower/Learner, but it leaves that committed vote in place. So after `startup()`:
> 
>     * `server_state == Follower`, `leader_ref() == None`
> 
>     * yet `state.is_leader(&id)` still returns `true` (the committed self-vote is untouched)
> 
>     * and storage still holds that committed vote
> 
> 
> That's a "follower" that `is_leader()` reports as a leader — an in-memory contradiction — on top of a memory/storage divergence. The new test actually shows it: the committed vote is left as-is and `take_commands()` is empty, so nothing reconciles or persists the change.

Thanks for this explanation, I've updated the test to explicitly assert that `state.is_leader(&id)` is `false` (which makes the test fail for now).

> So emulating standard Raft here means more than skipping the leader-state restore — we also have to reconcile the persisted leader info. The direction I lean toward: when the node reverts to follower instead of restoring leadership, demote the committed vote to uncommitted (keep the term, drop the committed bit so `is_leader()` goes false) and persist it with a `SaveVote`, leaving a clean follower at the same term.

I'll look a bit more deeply into the code to see how to do this and maybe try and think of any alternatives. Without knowing too much about how the internals work, I do worry that reverting the state of a vote from committed to uncommitted violates the irrevocable guarantee of consensus, that is once a decision is committed then it will never be uncommitted. Maybe there's just an additional check we can add to the `is_leader` check, something like `is_leader = is_leading && vote.is_committed() && vote.term > minimum_leading_term`, where `minimum_leading_term = if enable_leader_restore { 0 } else { current_term_at_startup }`.

> Let me think through the edge cases (lease handling, which term we keep) before we settle the exact handling.

Do you think it's worth making sure there's some turmoil test coverage for this new config? If so, I'll look into that.

**Comment by jkosh44** [comment]
Ok, just pushed another commit to "uncommit" the vote on the restarted leader.

File: tests-turmoil/src/bin/fuzz.rs [hunk 2]
```diff
@@ -72,6 +74,7 @@ impl DerivedConfig {
         let fail_rate = rng.gen_range(0.0..0.002);
         let election_timeout_max = election_timeout_min + rng.gen_range(100..500);
         let enable_chaos = rng.gen_bool(0.8);
+        let enable_leader_restore = rng.gen_bool(0.5);
```
Comment (jkosh44) [inline tests-turmoil/src/bin/fuzz.rs:76]: Let me know if you'd prefer another value here.

**Comment by jkosh44** [comment]
Just pushed another commit with the turmoil tests updated. It should be ready for another review.

**Review by drmingdrmer — CHANGES_REQUESTED** [review]
> @drmingdrmer made 1 comment.
> *[Reviewable](https://reviewable.io/reviews/databendlabs/openraft/1771)* status: 0 of 5 files reviewed, 3 unresolved discussions (waiting on jkosh44).

___
*[`openraft/src/engine/engine_impl.rs` line 154 at r5](https://reviewable.io/reviews/databendlabs/openraft/1771#-OurZLCVDk_D7K7pi38e:-OurZLCVDk_D7K7pi38f:b5vl1h3) ([raw file](https://github.com/databendlabs/openraft/blob/c21836722c1f3277cef6e7e32ba495da9265a3d9/openraft/src/engine/engine_impl.rs#L154)):*
> ```XML
>                 let uncommitted: VoteOf<C> = self.state.vote.to_non_committed().into_vote();
>                 self.state.vote.update(C::now(), Duration::default(), uncommitted.clone());
>                 self.output.push_command(Command::SaveVote {vote: uncommitted});
> ```

non-committed vote is smaller than the committed one. Saving it means to revert the `vote` value in storage. But `vote` value should be monotonically increasing. This change breaks this invariant. It should be done in some other way.



_Code quote:_
```XML
                let uncommitted: VoteOf<C> = self.state.vote.to_non_committed().into_vote();
                self.state.vote.update(C::now(), Duration::default(), uncommitted.clone());
                self.output.push_command(Command::SaveVote {vote: uncommitted});
```

**Comment by jkosh44** [comment]
> > @drmingdrmer made 1 comment.
> > _[Reviewable](https://reviewable.io/reviews/databendlabs/openraft/1771)_ status: 0 of 5 files reviewed, 3 unresolved discussions (waiting on jkosh44).
> 
> _[`openraft/src/engine/engine_impl.rs` line 154 at r5](https://reviewable.io/reviews/databendlabs/openraft/1771#-OurZLCVDk_D7K7pi38e:-OurZLCVDk_D7K7pi38f:b5vl1h3) ([raw file](https://github.com/databendlabs/openraft/blob/c21836722c1f3277cef6e7e32ba495da9265a3d9/openraft/src/engine/engine_impl.rs#L154)):_
> 
> > ```
> >                 let uncommitted: VoteOf<C> = self.state.vote.to_non_committed().into_vote();
> >                 self.state.vote.update(C::now(), Duration::default(), uncommitted.clone());
> >                 self.output.push_command(Command::SaveVote {vote: uncommitted});
> > ```
> 
> non-committed vote is smaller than the committed one. Saving it means to revert the `vote` value in storage. But `vote` value should be monotonically increasing. This change breaks this invariant. It should be done in some other way.
> 
> _Code quote:_
> 
> ```
>                 let uncommitted: VoteOf<C> = self.state.vote.to_non_committed().into_vote();
>                 self.state.vote.update(C::now(), Duration::default(), uncommitted.clone());
>                 self.output.push_command(Command::SaveVote {vote: uncommitted});
> ```

So it sounds like your earlier suggestion of flipping the committed flag does not work, am I understanding that correctly?

> The direction I lean toward: when the node reverts to follower instead of restoring leadership, demote the committed vote to uncommitted (keep the term, drop the committed bit so is_leader() goes false) and persist it with a SaveVote, leaving a clean follower at the same term.

One idea is to increment the term and flip the commit flag. I don't think that really makes sense though. Then we're just inventing a vote that never happened.

What do you think about my earlier suggestion?

> Maybe there's just an additional check we can add to the `is_leader` check, something like `is_leader = is_leading && vote.is_committed() && vote.term > minimum_leading_term`, where `minimum_leading_term = if enable_leader_restore { 0 } else { current_term_at_startup }`.

I can prototype something, I just want to make sure it sounds reasonable to you first.

**Comment by drmingdrmer** [comment]
AFAIK the most proper way is convert the in-memory vote from committed to non-committed. but do not revert the vote value in the storage. This introduce a in-consistency. But this inconsistency should not break anything I hope. Because there is not in-the-mid value between the non-committed and committed:  there is not any value `x`  so that `(term, node_id, non-committed) < x < (term, node_id, committed)`. And once the `vote` changes, the change should just automatically fix this inconsistency.

This is so far what I get.

**Comment by drmingdrmer** [comment]
@jkosh44 To answer your questions directly: flipping the committed flag is right — **but only in memory; the `SaveVote` is the bug**. `IOState` is initialized from the storage vote (committed), so saving the demoted (smaller) vote violates IO monotonicity and panics in debug builds. Repro: add `enable_leader_restore: Some(false)` to `t50_single_leader_restart_re_apply_logs`:

```text
panicked at openraft/src/raft_state/io_state/io_progress.rs:
expect submitted:(by:<T1-N0:Q>, None) < new_submitted:(<T1-N0:->)
```

The engine unit test can't catch this — it inspects commands but never executes them.

Why in-memory-only demotion is safe:

- Votes compare by `leader_id` first, then `committed`, so nothing orders between `(T, N, non-committed)` and `(T, N, committed)`. The demotion changes no grant/reject decision — only leadership derivation (`is_leader()`, `forward_to_leader()`, lease gate).
- Any vote greater than the demoted one is also `>=` the stored committed one, so every future `SaveVote` keeps storage (and IO progress) monotonic — the first vote change heals the divergence. No `SaveVote` is needed at startup.
- The resulting state — non-committed self-vote, Follower — is exactly a candidate that crashed after persisting its self-vote (see `test_startup_candidate_becomes_follower`). No new state space.

On irrevocability: the committed flag on a vote is a cached fact that leadership was quorum-granted, not a consensus decision on data — committed log entries are untouched. Standard Raft doesn't persist this fact at all; ignoring it at runtime *is* standard-Raft behavior.

I'd skip the `minimum_leading_term` guard: it makes every `vote.is_committed()` call site handle a new "committed but not leading" state — e.g. `forward_to_leader()` would keep pointing clients at this node while it refuses to lead.

I'll update the PR directly and push the changes to this branch:

1. `startup()`: drop the `push_command(Command::SaveVote ...)` line; keep only the in-memory update, with a comment on why storage keeps the committed vote.
2. Unit test: expect no commands; assert the in-memory vote is demoted.
3. A `tests/tests/life_cycle/` test: restart leader with flag off → Follower → re-elected at term+1 (the new term adds a noop entry, so expected log index +1).

I've verified these locally: all 475 lib tests pass, and the restarted leader re-elects at term 2 and re-applies all logs.

File: tests-turmoil/src/bin/fuzz.rs (hunk 2 above)
Comment (drmingdrmer) [reply tests-turmoil/src/bin/fuzz.rs:76]: it's ok with 0.5

**Review by drmingdrmer — APPROVED** [review]
> @drmingdrmer reviewed 8 files and all commit messages, made 1 comment, and resolved 3 discussions.
> *[Reviewable](https://reviewable.io/reviews/databendlabs/openraft/1771)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on jkosh44).

**Comment by jkosh44** [comment]
@drmingdrmer Thanks for the reviews and updates!
